### PR TITLE
Replicas: Always sort replicas by priority; Fix #4546

### LINKED ISSUE #4546: The replica sorting ignores the protocol priority
Motivation
----------

As reported by @rodwalker, the Rucio client does not necessarily honour the read_wan protocol priority.

Going through the implementation, in one example the value of `dictreplica` is:
```
{
  'srm://gse1.physik.uni-mainz.de:8444/srm/managerv2?SFN=/atlas/atlaslocalgroupdisk/rucio/mc15_13TeV/f7/e8/DAOD_TRUTH1.09608067._000002.pool.root.1': ('wan', 2, 'MAINZ_LOCALGROUPDISK', False),
  'davs://gdav1.physik.uni-mainz.de:8443/atlas/atlaslocalgroupdisk/rucio/mc15_13TeV/f7/e8/DAOD_TRUTH1.09608067._000002.pool.root.1': ('wan', 1, 'MAINZ_LOCALGROUPDISK', False)
}
```
but `sort_geoip` returns:
```
[
  'srm://gse1.physik.uni-mainz.de:8444/srm/managerv2?SFN=/atlas/atlaslocalgroupdisk/rucio/mc15_13TeV/f7/e8/DAOD_TRUTH1.09608067._000002.pool.root.1',
  'davs://gdav1.physik.uni-mainz.de:8443/atlas/atlaslocalgroupdisk/rucio/mc15_13TeV/f7/e8/DAOD_TRUTH1.09608067._000002.pool.root.1'
]
```
Similarly, with no sorting or with the yet-to-be-impemented methods, the order of the replicas is how Python returns `dictreplicas.keys()`.

Modification
------------

### PR DESCRIPTION
**Pull request by bziemons**
Add sorting by priority before doing any additional sort in `sort_replicas`. Resolves #4546

### CODE REVIEW

**Comment by dchristidis** [comment]
Should we maybe consider formalising the content of `dictreplicas`? Just by reading this file it’s impossible to guess what the values are.

Unless this is supposed to be a quick fix while more substantial work is underway?

**Comment by bziemons** [comment]
> Unless this is supposed to be a quick fix while more substantial work is underway?

exactly that. I was about to refactor that `dictreplicas` to something typed and checked, but I got told this is more time critical ;) Since the redirect endpoint also uses this `sort_replicas` function, this needs a bit of work and testing.

File: lib/rucio/core/replica_sorter.py [hunk 1]
```diff
@@ -179,9 +180,19 @@ def sort_replicas(dictreplica: "Dict", client_location: "Dict", selection: "Opti
     :param default: the default sorting algorithm (random, if not defined).
     :returns: the keys of dictreplica in a sorted list.
     """
+    if len(dictreplica) == 0:
+        return []
     if not selection:
         selection = 'geoip'
 
+    items = [(key, value) for key, value in dictreplica.items()]
```
Comment (dchristidis) [inline lib/rucio/core/replica_sorter.py:188]: Is this any different to `list(dictreplica.items())`?

File: lib/rucio/web/rest/flaskapi/v1/replicas.py [hunk 2]
```diff
@@ -134,12 +134,12 @@ def generate(vo):
                         yield f'  <glfn name="/atlas/rucio/{rfile["scope"]}:{rfile["name"]}">'
                         yield '</glfn>\n'
 
-                        idx = 0
+                        idx = 1
                         for replica in replicas:
```
Comment (dchristidis) [inline lib/rucio/web/rest/flaskapi/v1/replicas.py]: May I suggest `for idx, replica in enumerate(replicas, start=1)`?

**Review by dchristidis — DISMISSED** [review]

File: lib/rucio/core/replica_sorter.py (hunk 1 above)
Comment (bziemons) [reply lib/rucio/core/replica_sorter.py:188]: good question, have to test that :)

File: lib/rucio/web/rest/flaskapi/v1/replicas.py (hunk 2 above)
Comment (bziemons) [reply lib/rucio/web/rest/flaskapi/v1/replicas.py]: of course

File: lib/rucio/core/replica_sorter.py (hunk 1 above)
Comment (bziemons) [reply lib/rucio/core/replica_sorter.py:188]: while the standard documentation mentions that `dict.items()` *should* return an iterable of tuples, I'd rather be on the safe side here and make the tuples explicit.

**Review by dchristidis — APPROVED** [review]

**Review by bari12 — APPROVED** [review]

**Pull request merged**